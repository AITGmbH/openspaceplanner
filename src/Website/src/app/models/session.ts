import { Room } from "./room";
import { Slot } from "./slot";
import { Topic } from "./topic";

export class Session {
  public id: number;
  public name: string;
  public topics: Topic[];
  public rooms: Room[];
  public slots: Slot[];
  public createdAt: string;
  public freeForAll: boolean;
  public votingEnabled: boolean;
  public attendanceEnabled: boolean;
  public teamsAnnouncementsEnabled: boolean;

  constructor() {
    this.topics = [];
    this.slots = [];
    this.rooms = [];
  }
}
