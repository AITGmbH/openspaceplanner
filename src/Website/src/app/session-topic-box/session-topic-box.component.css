.topic {
  font-size: 0.9rem;
  width: 330px;
  max-width: 330px;
  float: left;
  margin: 5px;
}

.topic-indicators {
  position: relative;
  top: -20px;
  float: right;
}

.topic-content {
  position: relative;
}

.topic-header {
  height: 85px;
}

.topic-title {
  overflow: hidden;
  max-height: 48px;
}

.topic-owner {
  overflow: hidden;
  height: 24px;
  margin-bottom: 10px;
}

.topic-slots {
  position: absolute;
  background-color: #485fc7;
  color: white;
  font-size: 0.7rem;
  padding-top: 0.125rem;
  border-radius: 50%;
  width: 1.25rem;
  height: 1.25rem;
  text-align: center;
}

.content:not(:last-child) {
  margin-bottom: 0 !important;
}

.topic-description {
  height: 100%;
  margin-bottom: 20px;
}

.topic-description-expander {
  position: absolute;
  right: 0;
}

.topic-edit-button {
  position: absolute;
  right: 23px;
}

.topic-additional-informations {
  position: relative;
  top: 10px;
  width: 85%;
}

.topic-additional-information {
  float: left;
  width: 50%;
  padding-bottom: 5px;
}

.topic-additional-information div {
  display: inline;
}

.topic-additional-information>span~div {
  margin-left: 5px;
}

.topic-additional-information>span~span {
  margin-left: 5px;
}

.topic-buttons {
  clear: both;
  position: relative;
  bottom: 15px;
}

.topic-footer {
  width: 100%;
}

.topic-attendees span,
a {
  margin: 0 5px;
}

.topic-attendees-value {
  margin-right: 5px;
}

.topic-attendees-value-box {
  width: 20px;
}

.topic-ratings-value~i {
  color: #ffd700;
}

.topic-content small {
  top: 2px;
  position: relative;
}

.topic-ratings-stars {
  position: relative;
  left: -20px;
  width: calc(100% + 41px);
  top: 20px;
  border-top: 1px solid rgba(10, 10, 10, 0.1);
  clear: both;
}

.topic-ratings-stars>input {
  display: none;
}

.topic-ratings-stars>label:before {
  margin: 5px;
  font-size: 1.25em;
  font-family: FontAwesome;
  display: inline-block;
  content: "\f005";
}

.topic-ratings-stars>label {
  width: 25%;
  text-align: center;
  float: right;
  font-size: 14pt;
  border-right: 1px solid rgba(10, 10, 10, 0.1);
  color: #ddd;
}

.topic-ratings-stars>input:checked~label,
.topic-ratings-stars:not(:checked)>label:hover,
.topic-ratings-stars:not(:checked)>label:hover~label {
  color: #ffd700;
}

.topic-ratings-stars>input:checked+label:hover,
.topic-ratings-stars>input:checked~label:hover,
.topic-ratings-stars>label:hover~input:checked~label,
.topic-ratings-stars>input:checked~label:hover~label {
  color: #ffed85;
}

.room-seats-conflict {
  background-color: orange;
}

.same-owner-slot-conflict {
  background-color: rgb(255, 94, 94);
}

@media screen and (min-width: 1024px) {
  .topic {
    width: 100%;
    max-width: 450px;
    overflow: hidden;
  }
}

.topic-rating-tooltip {
  text-align: left;
  width: 100px;
}

.topic-error {
  color: orange;
  margin-right: 5px;
}

.topic-indicators span+span {
  margin-left: 5px;
}

.topic-indicators .fa-heart {
  color: red;
}

.topic-ratings-value {
  margin-left: 5px;
}