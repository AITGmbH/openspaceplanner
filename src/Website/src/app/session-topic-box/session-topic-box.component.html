<div class="topic box" [class.draggable]="enableDrag" id="{{ topic.id }}" appEditButton [class.topic-favorite]="isFavorite">
  <div class="topic-indicators">
    <span class="fa fa-comments" *ngIf="hasComments"></span>
    <a (click)="favorite()"><i class="fa fa-heart" *ngIf="isFavorite"></i></a>
    <span class="topic-slots" *ngIf="topic.slots > 1">{{ topic.slots }}</span>
  </div>

  <div class="content topic-content" #content>
    <div class="topic-header">
      <div class="topic-title">
        <span [hidden]="!hasError">
          <span class="fa fa-exclamation-triangle topic-error" #errorElement></span>
        </span>
        <div #errorTooltip>{{ errorText }}</div>
        <strong>{{ topic.name }}</strong>
      </div>

      <div class="topic-owner">
        <small>{{ topic.owner }}</small>
      </div>
    </div>

    <p class="topic-description" [hidden]="!description.dataset.expanded || description.dataset.expanded == 'false'" #description>
      <i *ngIf="!topic.description">no description</i>
      {{ topic.description }}
    </p>
  </div>

  <div class="topic-additional-informations">
    <div>
      <div class="topic-additional-information" *ngIf="session.attendanceEnabled">
        <span class="fa fa-users"></span>

        <div *ngIf="session.freeForAll" class="topic-attendees">
          <a (click)="increaseAttendees()">
            <i class="fa fa-plus" aria-hidden="true"></i>
          </a>

          <span class="topic-attendees-value" (click)="toggleAttendeesInput()" [hidden]="showAttendeesInput">{{ topic.attendees.length }}</span>
          <input
            type="text"
            class="topic-attendees-value-box"
            placeholder="1"
            #topicAttendees
            [hidden]="!showAttendeesInput"
            [value]="topic.attendees.length"
            (keyup.enter)="updateAttendees(topicAttendees.value)"
            (blur)="updateAttendees(topicAttendees.value)"
          />

          <a (click)="decreaseAttendees()">
            <i class="fa fa-minus" aria-hidden="true"></i>
          </a>
        </div>

        <div *ngIf="!session.freeForAll" class="topic-attendees">
          <span class="topic-attendees-value">{{ topic.attendees.length }}</span>

          <a (click)="increaseAttendees()" *ngIf="!isAttendee()">
            <i class="fa fa-plus" aria-hidden="true"></i>
          </a>

          <a (click)="decreaseAttendees()" *ngIf="isAttendee()">
            <i class="fa fa-minus" aria-hidden="true"></i>
          </a>
        </div>
      </div>

      <div class="topic-additional-information">
        <div #ratingElement>
          <span class="fa fa-star"></span>
          <span class="topic-ratings-value">{{ rating.mean }}</span>

          <div style="display: block" class="topic-rating-tooltip" #ratingTooltip>
            <div style="display: block">
              <span>Min:</span>
              <span class="pull-right">{{ rating.min }}</span>
            </div>
            <div style="display: block">
              <span>Median:</span>
              <span class="pull-right">{{ rating.median }}</span>
            </div>
            <div style="display: block">
              <span>Mean:</span>
              <span class="pull-right">{{ rating.mean }}</span>
            </div>
            <div style="display: block">
              <span>Max:</span>
              <span class="pull-right">{{ rating.max }}</span>
            </div>
            <div style="display: block">
              <span>Count:</span>
              <span class="pull-right">{{ rating.count }}</span>
            </div>
          </div>
        </div>
      </div>
    </div>

    <div *ngIf="showAdditionalInformation">
      <div class="topic-additional-information" *ngIf="slot != null">
        <span class="fa fa-clock-o"></span>
        <span>{{ slot.name }} {{ slot.time != null ? "(" + slot.time + ")" : "" }}</span>

        <div style="display: block" *ngIf="slotEnd != null">
          <span style="margin-left: 0.8rem">{{ slotEnd.name }} {{ slotEnd.time != null ? "(" + slotEnd.time + ")" : "" }}</span>
        </div>
      </div>

      <div class="topic-additional-information" *ngIf="room != null">
        <span class="fa fa-home"></span>
        <span>{{ room.name }}</span>
      </div>
    </div>
  </div>

  <div class="topic-buttons">
    <p class="topic-description-expander">
      <a (click)="expandDescription(content, description)" *ngIf="!description.dataset.expanded || description.dataset.expanded == 'false'">
        <i class="fa fa-expand"></i>
      </a>
      <a (click)="collapseDescription(content, description)" *ngIf="description.dataset.expanded == 'true'">
        <i class="fa fa-compress"></i>
      </a>
    </p>

    <p class="topic-edit-button">
      <a (click)="edit.emit({ event: $event, topic })" *ngIf="!showAdditionalInformation">
        <i class="fa fa-edit edit-button"></i>
      </a>

      <a (click)="favorite()">
        <i class="fa fa-heart edit-button" [class.fa-heart]="!isFavorite" [class.fa-heart-o]="isFavorite"></i>
      </a>
    </p>
  </div>

  <div class="topic-ratings-stars" *ngIf="session.votingEnabled">
    <input type="radio" id="star4-{{ topic.id }}" name="rating-{{ topic.id }}" value="4" (click)="ratingChange(4)" [checked]="getRating() == 4" />
    <label for="star4-{{ topic.id }}"></label>

    <input type="radio" id="star3-{{ topic.id }}" name="rating-{{ topic.id }}" value="3" (click)="ratingChange(3)" [checked]="getRating() == 3" />
    <label for="star3-{{ topic.id }}"></label>

    <input type="radio" id="star2-{{ topic.id }}" name="rating-{{ topic.id }}" value="2" (click)="ratingChange(2)" [checked]="getRating() == 2" />
    <label for="star2-{{ topic.id }}"></label>

    <input type="radio" id="star1-{{ topic.id }}" name="rating-{{ topic.id }}" value="1" (click)="ratingChange(1)" [checked]="getRating() == 1" />
    <label for="star1-{{ topic.id }}"></label>
  </div>
</div>
