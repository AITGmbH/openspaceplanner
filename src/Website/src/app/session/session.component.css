.table-cell:first-child {
  width: 180px;
}

.table-row {
  width: 100%;
}

tbody .table-row {
  height: 150px;
}

.session-table {
  margin-top: 20px;
  table-layout: fixed;
}

.session-table td,
.session-table th {
  border-width: 0 1px 1px 0;
}

.topics-unassigned {
  height: 150px;
}

.is-floating-action {
  border-radius: 50%;
  width: 52px;
  height: 52px;
  position: fixed;
  bottom: 15px;
  right: 15px;
  box-shadow: 0px 0px 3px rgba(0, 0, 0, 0.5), 3px 3px 3px rgba(0, 0, 0, 0.25);
}

.is-floating-button-submenu {
  position: fixed;
  bottom: 70px;
  right: 15px;
  color: #000;
}

.is-floating-button {
  border-radius: 50%;
  width: 52px;
  height: 52px;
  box-shadow: 0px 0px 3px rgba(0, 0, 0, 0.5), 3px 3px 3px rgba(0, 0, 0, 0.25);
  margin-bottom: 20px;
}

.is-floating-action,
.is-floating-button-submenu>.fa {
  font-size: 16px;
}

.is-floating-button-submenu li {
  background-color: #209cee;
  border-radius: 50%;
  width: 52px;
  height: 52px;
  margin-bottom: 20px;
  color: #fff;
  text-align: center;
}

.is-floating-button-submenu li .fa {
  position: relative;
  text-align: center;
  width: 100%;
}

.is-floating-button-submenu li div {
  position: relative;
  left: -330px;
  color: #000;
  text-align: right;
  width: 300px;
  top: -110%;
}

h1 .edit-button {
  margin-left: 10px;
}

.title {
  float: left;
}

.calendar-button {
  float: right;
}

.slot-not-planable {
  background-color: #ffe6ce;
}