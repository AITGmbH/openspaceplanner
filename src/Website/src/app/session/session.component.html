<div class="container is-loading" *ngIf="isLoading$ | async">Loading</div>
<h2 class="subtitle" *ngIf="session == null && !(isLoading$ | async)">Session not found.</h2>

<div class="container" *ngIf="session != null && !(isLoading$ | async)">
  <h1 class="title" (dblclick)="showModal($event, 'session', session)" appEditButton>
    {{ session.name }}
    <a (click)="showModal($event, 'session', session)">
      <span class="fa fa-edit edit-button"></span>
    </a>
  </h1>

  <a class="calendar-button" href="{{ calendarLink }}">
    <span class="fa fa-calendar fa-2x"></span>
  </a>

  <table class="table session-table is-striped is-fullwidth is-hoverable">
    <thead>
      <tr class="table-row">
        <td class="table-cell"></td>

        <td *ngFor="let room of rooms" class="table-cell" id="{{ room.id }}" (dblclick)="showModal($event, 'room', room)" appEditButton>
          <span>{{ room.name }}</span>
          <a (click)="showModal($event, 'room', room)">
            <span class="fa fa-edit pull-right edit-button"></span>
          </a>
          <br />
          ({{ room.seats }} seats)
        </td>
      </tr>
    </thead>

    <tbody>
      <tr class="table-row" *ngFor="let slot of slots" id="{{ slot.id }}">
        <td class="table-cell" (dblclick)="showModal($event, 'slot', slot)" appEditButton>
          <span>{{ slot.name }}</span>
          <a (click)="showModal($event, 'slot', slot)">
            <span class="fa fa-edit pull-right edit-button"></span>
          </a>
          <br />
          {{ slot.time }}
        </td>

        <ng-container *ngFor="let topic of topics[slot.id]">
          <td *ngIf="topic != null" class="table-cell topic-space" [class.slot-not-planable]="!slot.isPlanable" [attr.rowspan]="topic.slots" [attr.data-slot-id]="topic.slotId" [attr.data-room-id]="topic.roomId">
            <app-session-topic-box *ngIf="topic.id != null" [topic]="topic" (edit)="showModal($event.event, 'topic', $event.topic)" (dblclick)="showModal($event, 'topic', topic)"> </app-session-topic-box>
          </td>
        </ng-container>
      </tr>
    </tbody>
  </table>

  <div class="topics-unassigned dropable">
    <app-session-topic-box *ngFor="let topic of unassignedTopics" [topic]="topic" (dblclick)="showModal($event, 'topic', topic)" (edit)="showModal($event.event, 'topic', $event.topic)"></app-session-topic-box>
  </div>
</div>

<button (click)="toggleFloatingActionButton($event)" class="button is-info is-floating-action" #floatingActionButton>
  <span class="fa fa-plus"></span>
</button>

<ul *ngIf="floatingActionButton.dataset.expanded == 'true'" class="is-floating-button-submenu">
  <li>
    <button (click)="showModal($event, 'topic', {})" class="button is-info is-floating-button">
      <span class="fa fa-calendar"></span>
    </button>
    <div>Add Topic</div>
  </li>
  <li>
    <button (click)="showModal($event, 'room', {})" class="button is-info is-floating-button">
      <span class="fa fa-home"></span>
    </button>
    <div>Add Room</div>
  </li>
  <li>
    <button (click)="showModal($event, 'slot', {})" class="button is-info is-floating-button">
      <span class="fa fa-clock-o"></span>
    </button>
    <div>Add Slot</div>
  </li>
</ul>

<ng-container *ngIf="modalShown$ | async as modalShown">
  <app-topic-modal [isShown]="modalShown['topic']" (close)="hideModal('topic')" [item]="modalShown['topic']"></app-topic-modal>

  <app-room-modal [isShown]="modalShown['room']" (close)="hideModal('room')" [item]="modalShown['room']"></app-room-modal>

  <app-slot-modal [isShown]="modalShown['slot']" (close)="hideModal('slot')" [item]="modalShown['slot']"></app-slot-modal>

  <app-session-modal [isShown]="modalShown['session']" (close)="hideModal('session')" [item]="modalShown['session']"></app-session-modal>
</ng-container>
