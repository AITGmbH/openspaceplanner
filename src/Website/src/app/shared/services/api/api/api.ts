export * from './config.service';
import { ConfigService } from './config.service';
export * from './sessionRooms.service';
import { SessionRoomsService } from './sessionRooms.service';
export * from './sessionSlots.service';
import { SessionSlotsService } from './sessionSlots.service';
export * from './sessionTopics.service';
import { SessionTopicsService } from './sessionTopics.service';
export * from './sessionTopicsAttendance.service';
import { SessionTopicsAttendanceService } from './sessionTopicsAttendance.service';
export * from './sessionTopicsFeedback.service';
import { SessionTopicsFeedbackService } from './sessionTopicsFeedback.service';
export * from './sessionTopicsRating.service';
import { SessionTopicsRatingService } from './sessionTopicsRating.service';
export * from './sessions.service';
import { SessionsService } from './sessions.service';
export const APIS = [ConfigService, SessionRoomsService, SessionSlotsService, SessionTopicsService, SessionTopicsAttendanceService, SessionTopicsFeedbackService, SessionTopicsRatingService, SessionsService];
