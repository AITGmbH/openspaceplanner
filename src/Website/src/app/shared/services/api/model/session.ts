/**
 * OpenSpace.WebApi
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 1.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { Slot } from './slot';
import { Topic } from './topic';
import { VotingOptions } from './votingOptions';
import { Room } from './room';


export interface Session { 
    id: number;
    name: string;
    ratingEnabled: boolean;
    freeForAll: boolean;
    attendanceEnabled: boolean;
    teamsAnnouncementsEnabled: boolean;
    votingOptions: VotingOptions;
    createdAt: string;
    rooms: Array<Room>;
    slots: Array<Slot>;
    topics: Array<Topic>;
}

