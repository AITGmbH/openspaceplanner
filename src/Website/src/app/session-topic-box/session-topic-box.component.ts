import {
  Component,
  ElementRef,
  EventEmitter,
  Input,
  OnInit,
  Output,
  ViewChild
} from "@angular/core";
import tippy from "tippy.js";
import { RatingStatistic } from '../models/ratingStatistic';
import { SessionService } from "../session/session.service";
import { Room, Slot, Topic } from '../shared/services/api';

@Component({
  selector: "app-session-topic-box",
  templateUrl: "./session-topic-box.component.html",
  styleUrls: ["session-topic-box.component.css"],
})
export class SessionTopicBoxComponent implements OnInit {
  public showAttendeesInput: boolean = false;

  @Input()
  public topic!: Topic;

  @Input()
  public showAdditionalInformation: boolean = false;

  @Input()
  public enableDrag = true;

  @Output()
  public edit = new EventEmitter<{ event: Event, topic: Topic }>();

  public errors: string[] = [];

  @ViewChild("ratingTooltip", { static: true }) ratingTooltip!: ElementRef;

  @ViewChild("ratingElement", { static: true }) ratingElement!: ElementRef;

  @ViewChild("errorTooltip", { static: true }) errorTooltip!: ElementRef;

  @ViewChild("errorElement", { static: true }) errorElement!: ElementRef;

  @ViewChild("topicAttendees", { static: false }) topicAttendees!: ElementRef;

  public get isFavorite() {
    return this.sessionService.sessionOptions?.topicsFavorite[this.topic.id] || false;
  }

  public get hasComments() {
    return (
      this.topic != null &&
      this.topic.feedback != null &&
      this.topic.feedback.length > 0
    );
  }

  public get rating() {
    return new RatingStatistic(this.topic.ratings);
  }

  public get session() {
    return this.sessionService.currentSession;
  }

  public get errorText() {
    if (this.errors == null || this.errors.length === 0) {
      return "";
    }

    return this.errors.join(" ");
  }

  public get hasError() {
    this.errors = [];

    const hasErrors = [
      this.hasSameOwnerSlotConflict(),
      this.hasRoomSeatsConflict(),
      this.hasRoomCapabilitiesConflict(),
    ];

    return hasErrors.some(hasError => hasError);
  }

  public get slot() {
    if (this.session == null) {
      return null;
    }

    return this.session.slots.find((slot: Slot) => slot.id === this.topic.slotId);
  }

  public get slotEnd() {
    if (this.session == null) {
      return null;
    }

    if (this.topic.slots <= 1) {
      return null;
    }

    const slots = this.sessionService.getSortedSlots(this.session.slots);
    const slotIndex = slots.findIndex((slot: Slot) => slot.id === this.topic.slotId);
    return slots[slotIndex + this.topic.slots - 1];
  }

  public get room() {
    if (this.session == null) {
      return null;
    }

    return this.session.rooms.find((room: Room) => room.id === this.topic.roomId);
  }

  constructor(private sessionService: SessionService) { }

  public ngOnInit() {
    tippy(this.ratingElement.nativeElement, {
      content: this.ratingTooltip.nativeElement,
      interactive: true,
      allowHTML: true,
      appendTo: () => document.body
    });

    tippy(this.errorElement.nativeElement, {
      content: this.errorTooltip.nativeElement,
      interactive: true,
      allowHTML: true,
      appendTo: () => document.body
    });
  }

  public favorite() {
    this.sessionService.sessionOptions.topicsFavorite[this.topic.id] = !this.isFavorite;
    this.sessionService.saveSessionOptions();
  }

  public hasRoomSeatsConflict() {
    if (
      this.topic == null ||
      this.topic.slotId == null ||
      this.topic.roomId == null
    ) {
      return false;
    }

    const room = this.sessionService.currentSession.rooms.find((room: Room) => room.id === this.topic.roomId);
    if (room == null) {
      return false;
    }

    const seats = room.seats ?? 0;
    const hasError = seats < this.topic.attendees.length;
    if (hasError) {
      this.errors.push("Too many attendees for the given room size.");
    }

    return hasError;
  }

  public hasSameOwnerSlotConflict() {
    if (
      this.topic?.slotId == null ||
      this.topic.roomId == null ||
      this.topic.owner == null
    ) {
      return false;
    }

    const slots = this.sessionService.getSortedSlots(this.sessionService.currentSession.slots);

    const slotsOfCurrentTopic = this.sessionService.getSlotsOfTopic(slots, this.topic);

    const slotsOfTopicsWithSameOwner =
      this.sessionService.currentSession.topics.filter((topic: Topic) => topic.owner != null && topic.id !== this.topic.id && topic.owner === this.topic.owner)
        .reduce((a: Slot[], topic: Topic) => [...a, ...this.sessionService.getSlotsOfTopic(slots, topic)], []);

    const hasError = slotsOfCurrentTopic.some(slot => slotsOfTopicsWithSameOwner.includes(slot));
    if (hasError) {
      this.errors.push("Owner with two or more topics in the same slot.");
    }

    return hasError;
  }

  public hasRoomCapabilitiesConflict() {
    if (
      this.topic == null ||
      this.topic.slotId == null ||
      this.topic.roomId == null
    ) {
      return false;
    }

    if (this.topic.demands == null || this.topic.demands.length == 0) {
      return false;
    }

    const room = this.sessionService.currentSession.rooms.find((room: Room) => room.id === this.topic.roomId);
    if (room == null) {
      return false;
    }

    const hasError = !this.topic.demands.every(
      (d) => room.capabilities.findIndex((c: string) => c == d) >= 0
    );
    if (hasError) {
      this.errors.push(
        "The room does not have the capabilities required by the topic."
      );
    }

    return hasError;
  }

  public isAttendee() {
    return this.sessionService.getTopicAttendance(this.topic.id);
  }

  public increaseAttendees() {
    this.sessionService.updateTopicAttendance(this.topic.id, true);
  }

  public decreaseAttendees() {
    this.sessionService.deleteTopicAttendance(this.topic.id);
  }

  public ratingChange(rating: number) {
    if (this.getRating() === rating) {
      // voting for the same is equal to deleting your vote
      this.sessionService.deleteTopicRating(this.topic.id);
    } else {
      this.sessionService.updateTopicRating(this.topic.id, rating);
    }
  }

  public getRating() {
    return this.sessionService.getTopicRating(this.topic.id);
  }

  public expandDescription(content: HTMLElement, description: HTMLElement) {
    description.dataset.expanded = "true";
    content.dataset.expanded = "true";
  }

  public collapseDescription(content: HTMLElement, description: HTMLElement) {
    description.dataset.expanded = "false";
    content.dataset.expanded = "false";
  }

  public updateAttendees(attendeesText: string) {
    if (!this.showAttendeesInput) {
      return;
    }

    const attendees = parseInt(attendeesText, 10);
    this.showAttendeesInput = false;

    if (this.topic.attendees.length === attendees) {
      return;
    }

    this.sessionService.updateTopicAttendances(this.topic.id, attendees);
  }

  public toggleAttendeesInput() {
    this.showAttendeesInput = true;
    setTimeout(() => this.topicAttendees.nativeElement.focus());
  }
}
