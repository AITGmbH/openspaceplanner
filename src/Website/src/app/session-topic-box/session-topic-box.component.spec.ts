import { TestBed } from "@angular/core/testing";
import { By } from "@angular/platform-browser";
import * as moq from "typemoq";
import { SessionService } from "../session/session.service";
import { Session, Topic } from '../shared/services/api';
import { SessionTopicBoxComponent } from "./session-topic-box.component";

describe("session topic box", () => {
  it("should set error message when there is one owner with two topics in the same slot", () => {
    const session = {} as Session;
    session.topics.push(<Topic>{ owner: "Test", slotId: "1", roomId: "1" });
    session.topics.push(<Topic>{ owner: "Test", slotId: "1", roomId: "2" });

    const sessionServiceMock = moq.Mock.ofType<SessionService>();
    sessionServiceMock.setup(s => s.currentSession).returns(() => session);

    const comp = new SessionTopicBoxComponent(sessionServiceMock.object);
    comp.topic = session.topics[0];

    expect(comp.hasError).toBeTruthy();
    expect(comp.errors[0]).toBe("Owner with two or more topics in the same slot.");
  });

  it("should display error message when there are errors", () => {
    const session = {} as Session;
    session.topics.push(<Topic>{ owner: "Test", slotId: "1", roomId: "1" });
    session.topics.push(<Topic>{ owner: "Test", slotId: "1", roomId: "2" });

    const sessionServiceMock = moq.Mock.ofType<SessionService>();
    sessionServiceMock.setup(s => s.currentSession).returns(() => session);

    TestBed.configureTestingModule({
      declarations: [
        SessionTopicBoxComponent
      ],
      providers: [
        { provide: SessionService, useFactory: () => sessionServiceMock.object }
      ]
    }).compileComponents();

    const fixture = TestBed.createComponent(SessionTopicBoxComponent);
    const debugElement = fixture.debugElement;
    const comp = fixture.componentInstance;
    comp.topic = session.topics[0];

    fixture.detectChanges();

    const errorContainerElement: HTMLElement
      = debugElement.query(By.css(".topic-error")).parent?.nativeElement;

    expect(errorContainerElement.style.display).toBe("");
  });

  it("should set error message when there is one owner with two topics in the same slot (snapshot)", () => {
    const session = {} as Session;
    session.topics.push(<Topic>{ owner: "Test", slotId: "1", roomId: "1" });
    session.topics.push(<Topic>{ owner: "Test", slotId: "1", roomId: "2" });

    const sessionServiceMock = moq.Mock.ofType<SessionService>();
    sessionServiceMock.setup(s => s.currentSession).returns(() => session);

    TestBed.configureTestingModule({
      declarations: [
        SessionTopicBoxComponent
      ],
      providers: [
        { provide: SessionService, useFactory: () => sessionServiceMock.object }
      ]
    }).compileComponents();

    const fixture = TestBed.createComponent(SessionTopicBoxComponent);
    const comp = fixture.componentInstance;
    comp.topic = session.topics[0];

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
  });
});
