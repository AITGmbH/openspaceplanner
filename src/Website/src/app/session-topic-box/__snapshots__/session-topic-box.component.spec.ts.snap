// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`session topic box should set error message when there is one owner with two topics in the same slot (snapshot) 1`] = `
<div
  id="root1"
>
  <div
    appeditbutton=""
    class="topic box draggable"
    id="1"
  >
    <div
      class="topic-indicators"
    >
      
      
      
    </div>
    <div
      class="content topic-content"
    >
      <div
        class="topic-header"
      >
        <div
          class="topic-title"
        >
          <span>
            <span
              aria-expanded="false"
              class="fa fa-exclamation-triangle topic-error"
            />
          </span>
          <strong>
            
          </strong>
        </div>
        <div
          class="topic-owner"
        >
          <small>
            Test
          </small>
        </div>
      </div>
      <p
        class="topic-description"
        hidden=""
      >
        <i>
          no description
        </i>
        
          
      </p>
    </div>
    <div
      class="topic-additional-informations"
    >
      <div>
        
        <div
          class="topic-additional-information"
        >
          <div
            aria-expanded="false"
          >
            <span
              class="fa fa-star"
            />
            <span
              class="topic-ratings-value"
            >
              -
            </span>
          </div>
        </div>
      </div>
      
    </div>
    <div
      class="topic-buttons"
    >
      <p
        class="topic-description-expander"
      >
        <a>
          <i
            class="fa fa-expand"
          />
        </a>
        
        
      </p>
      <p
        class="topic-edit-button"
      >
        <a>
          <i
            class="fa fa-edit edit-button"
          />
        </a>
        
        <a>
          <i
            class="fa fa-heart edit-button"
          />
        </a>
      </p>
    </div>
    
  </div>
</div>
`;
